evm_arithmetization/src/prover.rs

```diff
@@ -30,132 +30,148 @@ use starky::stark::Stark;
-        state.registers = RegistersState {
-            program_counter: state.registers.program_counter,
-            is_kernel: state.registers.is_kernel,
-            is_stack_top_read: false,
-            check_overflow: false,
-            ..registers_before
-        };
-
-        let ExtraSegmentData {
-            trimmed_inputs,
-            bignum_modmul_result_limbs,
-            rlp_prover_inputs,
-            withdrawal_prover_inputs,
-            trie_root_ptrs,
-            jumpdest_table,
-        } = extra_segment_data.clone();
-
-        state.inputs = trimmed_inputs;
-        state.bignum_modmul_result_limbs = bignum_modmul_result_limbs;
-        state.rlp_prover_inputs = rlp_prover_inputs;
-        state.withdrawal_prover_inputs = withdrawal_prover_inputs;
-        state.trie_root_ptrs = trie_root_ptrs;
-        state.jumpdest_table = jumpdest_table;
-        let mut shift_addr = MemoryAddress::new(0, Segment::ShiftTable, 0);
-        let mut shift_val = U256::one();
-
-        for _ in 0..256 {
-            memory_before.set(shift_addr, shift_val);
-            shift_addr.increment();
-            shift_val <<= 1;
-        }
 
-        let actual_mem_before = {
-            let mut res = vec![];
-            for (ctx_idx, ctx) in memory_before.contexts.iter().enumerate() {
-                for (segment_idx, segment) in ctx.segments.iter().enumerate() {
-                    for (virt, value) in segment.content.iter().enumerate() {
-                        if let &Some(val) = value {
-                            res.push((
-                                MemoryAddress {
-                                    context: ctx_idx,
-                                    segment: segment_idx,
-                                    virt,
-                                },
-                                val,
-                            ));
-                        }
+    let mut state = GenerationState::<F>::new(&inputs, &KERNEL.code)
+        .map_err(|err| anyhow!("Failed to parse all the initial prover inputs: {:?}", err))?;
+
+    state.inputs = segment_data.extra_data.trimmed_inputs.clone();
+    state
+        .bignum_modmul_result_limbs
+        .clone_from(&segment_data.extra_data.bignum_modmul_result_limbs);
+    state
+        .rlp_prover_inputs
+        .clone_from(&segment_data.extra_data.rlp_prover_inputs);
+    state
+        .withdrawal_prover_inputs
+        .clone_from(&segment_data.extra_data.withdrawal_prover_inputs);
+    state.trie_root_ptrs = segment_data.extra_data.trie_root_ptrs.clone();
+    state
+        .jumpdest_table
+        .clone_from(&segment_data.extra_data.jumpdest_table);
+
+    let mut shift_addr = MemoryAddress::new(0, Segment::ShiftTable, 0);
+    let mut shift_val = U256::one();
+
+    for _ in 0..256 {
+        segment_data.memory.set(shift_addr, shift_val);
+        shift_addr.increment();
+        shift_val <<= 1;
+    }
```

Comment:
Just for clarity, maybe have this block in a tiny `initialize_shift_table_segment` helper method or something?